### PR TITLE
Allow usage of sprite identifiers in helpers; better error message for bad sprite names

### PR DESCRIPTION
**Pull request by alanhogan**
Two changes that should have positive effects on ease-of-use:
1. You can use the helper functions with an identifier instead of the full path name. (Either works.)
   
   So this works:
   
   ``` scss
   sprite-width($my-map, 'warning'); // for sprite name 'icons/warning.png'
   ```
2. A new error message if you still can't get the sprite name right.
   
   > error in C function sprite-width: No sprite of name or identifier 'not-gonna-happen' exists.
   > Sprites are named using their original asset source paths, like 'icons/shruggie.png'. You can also use a sprite identifier, like 'shruggie', which is provided via sprite-identifier(), as an argument to helper functions.

### CODE REVIEW

**Comment by alanhogan** [comment]
Regarding the first change — it really cleans up the API to let devs write `@include my-sprites('warning')`, so the knowledge of the import path will never be anywhere except the one spot where the spritemap was created. I hope you like this PR!

**Comment by alanhogan** [comment]
(`npm test` is successful for me. I don’t know what Travis is on about. It looks like `master` has build errors, too.)

**Comment by alanhogan** [comment]
What do you think, @joannajw @chriseppstein ?

**Comment by alanhogan** [comment]
Bump!

**Comment by eoneill** [comment]
I'm actually not a fan of this. While this works for simple use cases and does introduce some cursory convenience, it can also introduce confusion and issues down the road.

I think having two ways to reference a file within a sprite map complicates the implementation and has the potential to confuse people.

Additionally, the identifier is currently the filename, not the path to the file, so the current proposal fails with something like...

```
project/
└── assets/
    └── images/
        ├── icons/
        │   └── star.png
        └── mobile-icons/
            └── star.png
```

In this case, the identifier for both `icons/star.png` and `mobile-icons/star.png` is `star`. So if you looked up by identifier, which reference would you get?

**Comment by alanhogan** [comment]
That’s right, in your example, the last `star.png` is the one that will (always) be returned for the identifier `star`. 

Points to consider:
1. It’s _already_ the case that these two images will share the same identifier, so only the last `star` becomes the `star` of `.icon-star` if you use the generated identifiers to construct class names (which is their major use case). (Technically `.icon-star` would set the background properties twice, but only the last one would win)
2. If this behavior is unexpected, I have a hard time imagining what a user of this package _would_ expect. Similarly, I don’t think this error would be too hard to debug, since it's the userland input which is causing the problem.
3. This can be worked around by just using the longer names for these sprites.

However, I am a big fan of giving people heads-up when a tool can sense a likely error. I could add a simple detection to see if the same identifier is generated twice in a spritemap, and warn if so.

Regarding two ways to refer to a sprite being confusing:
1. Sure, probably one way is better… but if we are to pick one, I heavily prefer the _short_ option. It’s not only how things worked in Compass, but it’s DRYer — to me, it feels really dirty to have to hard-code the location of these images all over your Sass, instead of just one place.
2. More confusing than having two options is _not_ having the option of using the identifier. As a user of this lib, I found it quite counter-intuitive. Consider: It’s called `identifier` but we later learn this doesn’t actually identify the sprite. It comes _from_ eyeglass-spriting but you can’t give it back _to_ eyeglass-spriting. 

Thoughts?

**Comment by eoneill** [comment]
I think part of the issue is that `identifier` is a misnomer in the current implementation. We don't use the `identifier` to identify the sprite. The path is the only way to identify a sprite.

`sprite-identifier` doesn't refer to the `uid` of the sprite, but rather the [CSS IDENT](https://developer.mozilla.org/en-US/docs/Web/CSS/custom-ident) of the sprite.

So either we fix `sprite-identifier` to return a truly unique identifier (e.g. `icons--star`, `mobile-icons--star`, etc), or we just accept that `sprite-identifier` really just returns the basename.

If we don't fix it, I don't want to support looking up a sprite by it's basename.

If we do fix it, you'd have to manually convert the path to an ident or defer to eyeglass to get the identifier (e.g. you can't just use the basename).

So your implementation would now look something like...

``` scss
$sprite-id: sprite-identifier($sprite-map, 'path/to/img.png');
@include sprite-position($sprite-map, $sprite-id);
```

at which point it's much easier to just use the path as the lookup in the first place.

The only place that `sprite-identifier` is currently useful is to output a classname, which we know is broken per #16.

My vote is to fix `sprite-identifier` to to return a unique identifier that complies with CSS IDENT and add a `sprite-basename` function to return the current behavior.

I'll defer to @chriseppstein on this one, as he's more familiar with how Compass worked and why those decisions were made.

**Comment by alanhogan** [comment]
> So either we fix `sprite-identifier` to return a truly unique identifier (e.g. `icons--star`, `mobile-icons--star`, etc), or we just accept that `sprite-identifier` really just returns the `basename`.

I think it's going to quite unusual indeed for someone to one to sprite multiple folders of images together (whether the base names are the same or not). I’ve only ever seen a ratio of one sprited image to one folder. 

(Consider your example. The point of using sprites is performance. Why would I want to force all my users to download one big image that contains both mobile and desktop icons, when I really only want to show them one set?)

I would hate to see the complication that would result by extending the scope of what this package does to the point where it has non-obvious mappings of image file to suggested CSS names.

> If we don't fix it, I don't want to support looking up a sprite by it's basename.

Why not? As a Sass author, this is a huge win, all upside, as far as I can tell! Except of course for the pathological case of identically named sprites, which is GIGO.

**Comment by chriseppstein** [comment]
> Why not? As a Sass author, this is a huge win, all upside, as far as I can tell!

My experience with compass is that by abstracting the name of the image that is being referenced in the sprite, it actually created a lot of confusion. So in eyeglass we specifically decided to use the asset's path to reference that file within the sprite. Yes, it's more verbose, but it's also a lot more clear and when people are searching CSS files for where a file they are about to remove is referenced, this approach is a lot more likely to get a hit.

> My vote is to fix sprite-identifier to to return a unique identifier that complies with CSS IDENT and add a sprite-basename function to return the current behavior.

We don't need to add `sprite-basename`. The filesystem module has a `fs-parse-filename($filename)` function that gives access to the base name.

The issue is that not everyone will even use the `sprite-identifier` function. So forcing people to think about identifiers or work around collisions sucks for them. In issue #4 we plan to add exceptions so that you can assign a sprite identifier to a filename -- or, using the filesystem module, those exceptions could be generated by a helper function.

I also think we could have an option to allow sprite identifiers to be auto-generated instead of based on the filename so that they will be guaranteed unique in the map but the user doesn't know what they will be ahead of time (See Sass's `unique-id()`).

I'd also be ok with giving an error if `sprite-identifier` is ever called for a file that doesn't have a unique identifier (thus forcing them to add an exception).

### TL;DR

I don't want to support this API change as part of eyeglass spriting, but it should be quite easy to build this API on top of our existing sprite map data structure in pure Sass and release it as an eyeglass module for those who prefer this approach.
